src/ggml-cuda/out-prod.cu

```diff
@@ -0,0 +1,59 @@
+#include "out-prod.cuh"
+#include "vendors/cuda.h"
+
+#include <cstdint>
+
+void ggml_cuda_out_prod(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
+    const ggml_tensor * src0 = dst->src[0];
+    const ggml_tensor * src1 = dst->src[1];
+
+    GGML_ASSERT(src0->type == GGML_TYPE_F32);
+    GGML_ASSERT(src1->type == GGML_TYPE_F32);
+    GGML_ASSERT(dst->type  == GGML_TYPE_F32);
+    GGML_ASSERT(ggml_is_contiguous(src0));
+    GGML_ASSERT(ggml_is_contiguous(dst));
+
+    const int64_t ne00 = src0->ne[0];
+    const int64_t ne01 = src0->ne[1];
+
+    const int64_t ne10 = src1->ne[0];
+    const int64_t ne11 = src1->ne[1];
+
+    GGML_ASSERT(ne01 == ne11);
+
+    const int64_t ne0 = dst->ne[0];
+    const int64_t ne1 = dst->ne[1];
+    const int64_t ne2 = dst->ne[2];
+    const int64_t ne3 = dst->ne[3];
+
+    GGML_ASSERT(ne0 == ne00);
+    GGML_ASSERT(ne1 == ne10);
+
+    GGML_ASSERT(ne2 == src0->ne[2]);
+    GGML_ASSERT(ne2 == src1->ne[2]);
+    GGML_ASSERT(ne3 == src0->ne[3]);
+    GGML_ASSERT(ne3 == src1->ne[3]);
+
+    const float * src0_d = (const float *) src0->data;
+    const float * src1_d = (const float *) src1->data;
+    float       *  dst_d = (float       *)  dst->data;
+
+    cudaStream_t   stream = ctx.stream();
+    cublasHandle_t handle = ctx.cublas_handle();
+
+    const float alpha = 1.0f;
+    const float beta = 0.0f;
+
+    GGML_ASSERT(ne2 == 1);
+    GGML_ASSERT(ne3 == 1);
+    CUBLAS_CHECK(cublasSetStream(handle, stream));
+
+    const cublasOperation_t src1_cublas_op = ggml_is_transposed(src1) ? CUBLAS_OP_N : CUBLAS_OP_T;
+    const int64_t           ldb            = ggml_is_transposed(src1) ? ne11        : ne10;
```

Comment:
I think `ldb = nb11/sizeof(float)` should work more generically. Otherwise it will fail silently with some types of views.

Reply:
I think that will result in the wrong logic. What I need here is to return a constant value for `ldb` regardless of whether or not `src1` was transposed, unless I'm missing something the value of `nb11/sizeof(float)` would change.

Reply:
Yes, you are right. I was thinking about strides, but didn't take into account that cublas already expects a transposed matrix. I suggest this change instead:

```diff
diff --git a/src/ggml-cuda/out-prod.cu b/src/ggml-cuda/out-prod.cu
index ee621d3..7807474 100644
--- a/src/ggml-cuda/out-prod.cu
+++ b/src/ggml-cuda/out-prod.cu
@@ -7,25 +7,15 @@ void ggml_cuda_out_prod(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
     const ggml_tensor * src0 = dst->src[0];
     const ggml_tensor * src1 = dst->src[1];

+    GGML_TENSOR_BINARY_OP_LOCALS
+
     GGML_ASSERT(src0->type == GGML_TYPE_F32);
     GGML_ASSERT(src1->type == GGML_TYPE_F32);
     GGML_ASSERT(dst->type  == GGML_TYPE_F32);
     GGML_ASSERT(ggml_is_contiguous(src0));
     GGML_ASSERT(ggml_is_contiguous(dst));

-    const int64_t ne00 = src0->ne[0];
-    const int64_t ne01 = src0->ne[1];
-
-    const int64_t ne10 = src1->ne[0];
-    const int64_t ne11 = src1->ne[1];
-
     GGML_ASSERT(ne01 == ne11);
-
-    const int64_t ne0 = dst->ne[0];
-    const int64_t ne1 = dst->ne[1];
-    const int64_t ne2 = dst->ne[2];
-    const int64_t ne3 = dst->ne[3];
-
     GGML_ASSERT(ne0 == ne00);
     GGML_ASSERT(ne1 == ne10);

@@ -48,8 +38,8 @@ void ggml_cuda_out_prod(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
     GGML_ASSERT(ne3 == 1);
     CUBLAS_CHECK(cublasSetStream(handle, stream));

-    const cublasOperation_t src1_cublas_op = ggml_is_transposed(src1) ? CUBLAS_OP_N : CUBLAS_OP_T;
-    const int64_t           ldb            = ggml_is_transposed(src1) ? ne11        : ne10;
+    const cublasOperation_t src1_cublas_op =  ggml_is_transposed(src1) ? CUBLAS_OP_N : CUBLAS_OP_T;
+    const int64_t           ldb            = (ggml_is_transposed(src1) ?        nb10 :        nb11)/sizeof(float);
     CUBLAS_CHECK(
         cublasSgemm(handle, CUBLAS_OP_N, src1_cublas_op,
                 ne0, ne1, ne01,
diff --git a/tests/test-backend-ops.cpp b/tests/test-backend-ops.cpp
index 729d4b2..6c4de06 100644
--- a/tests/test-backend-ops.cpp
+++ b/tests/test-backend-ops.cpp
@@ -1679,9 +1679,10 @@ struct test_out_prod : public test_case {
     const int64_t n;
     const int64_t k;
     const std::array<int64_t, 2> bs; // dims 3 and 4
+    const bool trans_b;

     std::string vars() override {
-        return VARS_TO_STR6(type_a, type_b, m, n, k, bs);
+        return VARS_TO_STR7(type_a, type_b, m, n, k, bs, trans_b);
     }

     double max_nmse_err() override {
@@ -1690,13 +1691,21 @@ struct test_out_prod : public test_case {

     test_out_prod(ggml_type type_a = GGML_TYPE_F32, ggml_type type_b = GGML_TYPE_F32,
             int64_t m = 32, int64_t n = 32, int64_t k = 32,
-            std::array<int64_t, 2> bs = {10, 10})
-        : type_a(type_a), type_b(type_b), m(m), n(n), k(k), bs(bs) {}
+            std::array<int64_t, 2> bs = {10, 10},
+            bool trans_b = false)
+        : type_a(type_a), type_b(type_b), m(m), n(n), k(k), bs(bs), trans_b(trans_b) {}

     ggml_tensor * build_graph(ggml_context * ctx) override {
         ggml_tensor * a = ggml_new_tensor_4d(ctx, type_a, m, k, bs[0], bs[1]);
-        ggml_tensor * b = ggml_new_tensor_4d(ctx, type_b, n, k, bs[0], bs[1]);
         ggml_set_name(a, "a");
+
+        ggml_tensor * b;
+        if (trans_b) {
+            b = ggml_new_tensor_4d(ctx, type_b, k, n, bs[0], bs[1]);
+            b = ggml_transpose(ctx, b);
+        } else {
+            b = ggml_new_tensor_4d(ctx, type_b, n, k, bs[0], bs[1]);
+        }
         ggml_set_name(b, "b");

         ggml_tensor * out = ggml_out_prod(ctx, a, b);
@@ -3432,6 +3441,7 @@ static bool test_backend(ggml_backend_t backend, test_mode mode, const char * op
             test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 1, 16, {10, 10}));

             test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 16, 16, { 1,  1}));
+            test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 16, 16, { 1,  1}, true));
             test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 16, 16, {10,  1}));
             test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 16, 16, {10,  1}));
             test_cases.emplace_back(new test_out_prod(type_a, type_b, 256, 16, 16, {10, 10}));
```

Reply:
Thanks, I also added an assert for the other stride being `sizeof(float)` since this is required by cuBLAS.